### PR TITLE
Parse arrows with correct precedence

### PR DESCRIPTION
**Pull request by bakkot**

### CODE REVIEW

File: src/main/java/com/shapesecurity/shift/es2016/utils/Either3.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	public Maybe<C> right() {
+		return this.tag == Tag.RIGHT ? Maybe.of((C) this.data) : Maybe.empty();
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) return true;
+		if (o == null || getClass() != o.getClass()) return false;
+		Either3<?, ?, ?> either3 = (Either3<?, ?, ?>) o;
+		return Objects.equals(data, either3.data) &&
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2016/utils/Either3.java]: Let's compare the tag first.

File: src/main/java/com/shapesecurity/shift/es2016/utils/Either3.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	@SuppressWarnings("unchecked")
+	@Nonnull
+	public Maybe<C> right() {
+		return this.tag == Tag.RIGHT ? Maybe.of((C) this.data) : Maybe.empty();
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) return true;
+		if (o == null || getClass() != o.getClass()) return false;
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2016/utils/Either3.java:107]: We typically just do `o instanceof Either3`.

**Review by michaelficarra — COMMENTED** [review]
I'd really like to see a comment somewhere that describes how the Either3s flow through the expression parser and what conditions are met when we have a Left, Middle, or Right. Not sure where to put it though.

File: src/main/java/com/shapesecurity/shift/es2016/parser/ErrorMessages.java [hunk 3]
```diff
@@ -54,6 +54,7 @@
     String INVALID_VAR_INIT_FOR_IN = "Invalid variable declaration in for-in statement";
     String INVALID_VAR_INIT_FOR_OF = "Invalid variable declaration in for-of statement";
     String ILLEGAL_PROPERTY = "Illegal property initializer";
+    String UNEXPECTED_ARROW = "ArrowExpressions may not appear in this position";
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/shift/es2016/parser/ErrorMessages.java]: These messages are written for humans.  Probably just

> Arrows may not appear in this position.

**Comment by bakkot** [comment]
> I'd really like to see a comment somewhere that describes how the Either3s flow through the expression parser and what conditions are met when we have a Left, Middle, or Right. Not sure where to put it though.

One option would be to use a specialized helper class with meaningful method names instead of `Either`.

**Comment by michaelficarra** [comment]
Name alone probably wouldn't be sufficient here. I don't want to see a name like `CouldBeAnExpressionOrAssignmentTargetButNotAnArrowHead`.

File: src/main/java/com/shapesecurity/shift/es2016/utils/Either3.java (hunk 2 above)
Comment (bakkot) [reply src/main/java/com/shapesecurity/shift/es2016/utils/Either3.java:107]: These methods were auto-generated by IntelliJ, which I assume knows what it's doing.

**Comment by bakkot** [comment]
Addressed comments.

**Review by michaelficarra — APPROVED** [review]

**Pull request merged**